clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/providers/TitusTargetGroupServerGroupProvider.java

```diff
@@ -48,77 +49,68 @@ public TitusTargetGroupServerGroupProvider(Cache cacheView) {
+
     for (CacheData serverGroup : applicationServerGroups) {
       if (serverGroup.getRelationships().containsKey(TARGET_GROUPS.ns)) {
         for (String targetGroup : serverGroup.getRelationships().get(TARGET_GROUPS.ns)) {
-          Map targetGroupDetails = com.netflix.spinnaker.clouddriver.aws.data.Keys.parse(targetGroup);
+          Map<String, String> targetGroupDetails = com.netflix.spinnaker.clouddriver.aws.data.Keys.parse(targetGroup);
+
           Set<LoadBalancerInstance> targetGroupInstances = new HashSet<>();
           if (serverGroup.getRelationships().containsKey(INSTANCES.ns)) {
             for (String instanceKey : serverGroup.getRelationships().get(INSTANCES.ns)) {
               Map instanceDetails = instances.get(instanceKey);
-              String healthKey = com.netflix.spinnaker.clouddriver.aws.data.Keys.getInstanceHealthKey(((Map) instanceDetails.get("task")).get("containerIp").toString(), targetGroupDetails.get("account").toString(), targetGroupDetails.get("region").toString(), "aws-load-balancer-v2-target-group-instance-health");
-              CacheData healthData = cacheView.get(HEALTH.ns, healthKey);
-              Map health = Collections.EMPTY_MAP;
-              try {
-                if (healthData != null
-                  && healthData.getAttributes().containsKey("targetGroups")
-                  && !((ArrayList) healthData.getAttributes().get("targetGroups")).isEmpty()) {
-                  Map targetGroupHealth = (Map) ((ArrayList) healthData.getAttributes().get("targetGroups")).stream().filter(tgh ->
-                    ((Map) tgh).get("targetGroupName").toString().equals(targetGroupDetails.get("targetGroup")
-                    )).findFirst().orElse(Collections.EMPTY_MAP);
-                  if (!targetGroupHealth.isEmpty()) {
-                    health = new HashMap<String, String>() {
-                      {
-                        put("targetGroupName", targetGroupHealth.get("targetGroupName").toString());
-                        put("state", targetGroupHealth.get("state").toString());
-                        if (targetGroupHealth.containsKey("reasonCode")) {
-                          put("reasonCode", targetGroupHealth.get("reasonCode").toString());
-                        }
-                        if (targetGroupHealth.containsKey("description")) {
-                          put("description", targetGroupHealth.get("description").toString());
-                        }
-                      }
-                    };
-                  }
-                }
-              } catch (Exception e) {
-                log.error("failed to load health for " + instanceKey, e);
+
+              Optional<LoadBalancerInstance> instance = getInstanceHealth(instanceKey, instanceDetails, targetGroupDetails);
+              if (instance.isPresent()) {
+                targetGroupInstances.add(instance.get());
+              } else {
+                log.error(format(
```

Comment:
don't really need `String.format()` ... `"Detected potentially .. (targetGroup: {}, serverGroup: {})"` should suffice?

Reply:
thinking out loud --- I wonder where else we could start looking for a signal that caches are potentially inconsistent in some way and emit a metric (short of a user telling us)

Reply:
This is a potential problem anywhere we're resolving relationship data between two types, right? If that assertion is correct, I'd tend to think the problem could pop up in quite a number of places, like in EC2 load balancer health checks.

Reply:
Yes, I believe that's true.

Something to keep on the back burner, it'd be nice to more readily detect when things are consistently inconsistent.